runners/prism/java/src/main/java/org/apache/beam/runners/prism/PrismLocator.java

```diff
@@ -0,0 +1,253 @@
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.hash.HashCode;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.hash.Hashing;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.io.ByteStreams;
+
+/**
+ * Locates a Prism executable based on a user's default operating system and architecture
+ * environment or a {@link PrismPipelineOptions#getPrismLocation()} override. Handles the download,
+ * unzip, {@link PosixFilePermissions}, as needed. For {@link #GITHUB_DOWNLOAD_PREFIX} sources,
+ * additionally performs a SHA512 verification.
+ */
+class PrismLocator {
+  static final String OS_NAME_PROPERTY = "os.name";
+  static final String ARCH_PROPERTY = "os.arch";
+  static final String USER_HOME_PROPERTY = "user.home";
+
+  private static final String ZIP_EXT = "zip";
+  private static final String SHA512_EXT = "sha512";
+  private static final ReleaseInfo RELEASE_INFO = ReleaseInfo.getReleaseInfo();
+  private static final String PRISM_BIN_PATH = ".apache_beam/cache/prism/bin";
+  private static final Set<PosixFilePermission> PERMS =
+      PosixFilePermissions.fromString("rwxr-xr-x");
+  private static final String GITHUB_DOWNLOAD_PREFIX =
+      "https://github.com/apache/beam/releases/download";
+  private static final String GITHUB_TAG_PREFIX = "https://github.com/apache/beam/releases/tag";
+
+  private final PrismPipelineOptions options;
+
+  PrismLocator(PrismPipelineOptions options) {
+    this.options = options;
+  }
+
+  /**
+   * Downloads and prepares a Prism executable for use with the {@link PrismRunner}, executed by the
+   * {@link PrismExecutor}. The returned {@link String} is the absolute path to the Prism
+   * executable.
+   */
+  String resolve() throws IOException {
+
+    String from =
+        String.format("%s/v%s/%s.zip", GITHUB_DOWNLOAD_PREFIX, getSDKVersion(), buildFileName());
+
+    if (!Strings.isNullOrEmpty(options.getPrismLocation())) {
+      checkArgument(
+          !options.getPrismLocation().startsWith(GITHUB_TAG_PREFIX),
+          "Provided --prismLocation URL is not an Apache Beam Github "
+              + "Release page URL or download URL: ",
+          from);
+
+      from = options.getPrismLocation();
+    }
+
+    String fromFileName = getNameWithoutExtension(from);
+    Path to = Paths.get(userHome(), PRISM_BIN_PATH, fromFileName);
+
+    if (Files.exists(to)) {
+      return to.toString();
+    }
+
+    createDirectoryIfNeeded(to);
+
+    if (from.startsWith("http")) {
+      String result = resolve(new URL(from), to);
+      checkState(Files.exists(to), "Resolved location does not exist: %s", result);
+      return result;
+    }
+
+    String result = resolve(Paths.get(from), to);
+    checkState(Files.exists(to), "Resolved location does not exist: %s", result);
+    return result;
+  }
+
+  private String resolve(URL from, Path to) throws IOException {
+    if (from.toString().startsWith(GITHUB_DOWNLOAD_PREFIX)) {
```

Comment:
Will this always be true? Should we just assert this instead?

Reply:
Not always.